lib/read/lib/Cardano/Wallet/Read/Block/Gen/Build.hs

```diff
@@ -0,0 +1,401 @@
+input i txid' = liftF $ Input i txid' ()
+
+output :: MonadFree TxBuild m => Address -> Lovelace -> m ()
+output addr val = liftF $ Output addr val ()
+
+data ChainBuild m pa a where
+    Tx :: TxF m b -> (TxId -> a) -> ChainBuild m pa a
+    TxP :: TxParameters -> (TxId -> a) -> ChainBuild m pa a
+    NewAddress :: pa -> (Address -> a) -> ChainBuild m pa a
+    Byron :: SlotNo -> a -> ChainBuild m pa a
+    Shelley :: SlotNo -> a -> ChainBuild m pa a
+    Allegra :: SlotNo -> a -> ChainBuild m pa a
+    Mary :: SlotNo -> a -> ChainBuild m pa a
+    Alonzo :: SlotNo -> a -> ChainBuild m pa a
+    Babbage :: SlotNo -> a -> ChainBuild m pa a
+    Conway :: SlotNo -> a -> ChainBuild m pa a
+
+deriving instance Functor (ChainBuild pa m)
+
+hoistChainBuild
+    :: Functor m
+    => (forall x. m x -> n x)
+    -> ChainBuild m pa a
+    -> ChainBuild n pa a
+hoistChainBuild f x = case x of
+    Tx txf g -> Tx (hoistFreeT f txf) g
+    TxP txp g -> TxP txp g
+    NewAddress k g -> NewAddress k g
+    Byron sn a -> Byron sn a
+    Shelley sn a -> Shelley sn a
+    Allegra sn a -> Allegra sn a
+    Mary sn a -> Mary sn a
+    Alonzo sn a -> Alonzo sn a
+    Babbage sn a -> Babbage sn a
+    Conway sn a -> Conway sn a
+
+type ChainF m pa = FreeT (ChainBuild m pa) m
+
+hoistChainF
+    :: Monad m
+    => (forall x. m x -> n x)
+    -> ChainF m pa a
+    -> ChainF n pa a
+hoistChainF f = hoistFreeT f . transFreeT (hoistChainBuild f)
+
+tx :: Monad m => TxF m a -> ChainF m pa TxId
+tx txf = liftF $ Tx txf id
+
+txP :: Monad m => TxParameters -> ChainF m pa TxId
+txP txp = liftF $ TxP txp id
+
+address :: Monad m => pa -> ChainF m pa Address
+address k = liftF $ NewAddress k id
+
+block
+    :: Monad m
+    => (SlotNo -> () -> ChainBuild m pa ())
+    -> Natural
+    -> ChainF m pa ()
+block era slot = liftF $ era (SlotNo slot) ()
+
+byron :: Monad m => Natural -> ChainF m pa ()
+byron = block Byron
+
+shelley :: Monad m => Natural -> ChainF m pa ()
+shelley = block Shelley
+
+allegra :: Monad m => Natural -> ChainF m pa ()
+allegra = block Allegra
+
+mary :: Monad m => Natural -> ChainF m pa ()
+mary = block Mary
+
+alonzo :: Monad m => Natural -> ChainF m pa ()
+alonzo = block Alonzo
+
+babbage :: Monad m => Natural -> ChainF m pa ()
+babbage = block Babbage
+
+conway :: Monad m => Natural -> ChainF m pa ()
+conway = block Conway
+
+--------------------------------------------------------------------------------
+-- ChainF interpreter
+--------------------------------------------------------------------------------
+
+-- | Streaming output for the interpreter
+type ChainFOutput e m = Stream (Of e) m
+
+-- | Streams of 'BlockParameters' as from the
+type BlockParametersStream m = ChainFOutput (EraValue BlockParameters) m
+
+type ConsensusBlocksStream m = ChainFOutput ConsensusBlock m
+
+-- | Generate blocks from a 'ChainF ()' value
+genChainFBlocks
+    :: Monad m
+    => (pa -> m Address)
+    -> ChainF m pa ()
+    -> ConsensusBlocksStream m ()
```

Comment:
```suggestion
    -> m [EraValue Block]
```

Two requests:

* As far as `Read` is concerned, the main abstraction is `Block`, not `ConsensusBlock`. Hence, to make this module self-contained, I would like to return `Block` here.
* I would like to avoid the use of an additional library `streaming` — every additional module and additional library adds to mental and import load and makes the abstraction less useable. The performance optimization by using streaming is premature, returning a list is perfectly fine. The function `interpretChainF` can be reformulated to return `m […]` as well without much change. Note that the concept of streaming only appears because we use a monad `m` in the first place — a pure function would return a list `[…]` which, thanks to lazy evaluation, has streaming built-in. Monads just don't compose very well compared to pure functions.

Reply:
returning `m [a]`  is producing a lazy list of `a` ? I would like to avoid this rabbit hole and interleave a monadic layer. But we can roll our own streaming layer . I will give it a go

Reply:
> returning `m [a]` is producing a lazy list of `a` ?

No, it's likely not producing a lazy list (though there are situations where it does, e.g. when `m = Identity`) — what I'm saying is that not producing a lazy list is perfectly fine, for the sake of a simpler interface.